codalab/worker_manager/worker_manager.py

```diff
@@ -0,0 +1,167 @@
+    and bundles are related - that logic is complex and is done by the usual
+    worker system, and we don't want to duplicate that for every possible
+    worker backend.
+
+    More specifically, a worker manager will monitor a job queue to see how
+    many worker jobs are running, and try to keep that between `min_workers`
+    and `max_workers`.  It will also monitor the staged bundles that satisfy a
+    certain `search` criterion.  If there are staged bundles then it will issue
+    a `start_worker_job()` call, provided some other conditions are met (e.g.,
+    don't start workers too fast).
+
+    The WorkerManager is all client-side code, so it can be customized as one
+    sees fit.
+
+    Notes:
+    - The worker manager is not visible via CodaLab (i.e., CodaLab has no
+      notion of a worker manager or what it's trying to do - all it sees is
+      bundles and workers).  One needs to monitor the AWS/Azure Batch system
+      separately.
+    - Resource handling is not currently supported.  Generally, the safe thing
+      is to create a separate queue for different resource needs and put the
+      burden of deciding on the user.
+    """
+
+    def __init__(self, args):
+        self.args = args
+        self.codalab_manager = CodaLabManager()
+        self.codalab_client = self.codalab_manager.client(args.server)
+        self.staged_uuids = []
+        self.last_worker_start_time = 0
+        logger.info('Started worker manager.')
+
+    def get_worker_jobs(self):
+        """Return a list of `WorkerJob`s."""
+        raise NotImplementedError
+
+    def start_worker_job(self):
+        """Start a new `WorkerJob`."""
+        raise NotImplementedError
+
+    def run_loop(self):
+        while True:
+            self.run_one_iteration()
+            if self.args.once:
+                break
+            logger.debug('Sleeping {} seconds'.format(self.args.sleep_time))
+            time.sleep(self.args.sleep_time)
+
+    def run_one_iteration(self):
+        # Get staged bundles
+        keywords = ['state=' + State.STAGED] + self.args.search
+        if self.args.worker_tag:
+            keywords.append('request_queue=tag=' + self.args.worker_tag)
+        bundles = self.codalab_client.fetch(
+            'bundles', params={'worksheet': None, 'keywords': keywords, 'include': ['owner']}
+        )
+        new_staged_uuids = [bundle['uuid'] for bundle in bundles]
+        old_staged_uuids = self.staged_uuids
+        # Bundles that were staged but now aren't
+        removed_uuids = [uuid for uuid in old_staged_uuids if uuid not in new_staged_uuids]
+        self.staged_uuids = new_staged_uuids
+        logger.info(
+            'Staged bundles [{}]: {}'.format(
+                ' '.join(keywords), ' '.join(self.staged_uuids) or '(none)'
+            )
+        )
+
+        # Get worker jobs
+        worker_jobs = self.get_worker_jobs()
+        pending_jobs, active_jobs = [], []
```

Comment:
Can we rename to `pending_worker_jobs` and `active_worker_jobs`?  `worker` is more important than `job` here, because `job` could mean bundle.